libcxx/include/__iterator/aliasing_iterator.h

```diff
@@ -102,7 +102,7 @@ struct __aliasing_iterator_wrapper {
 
     _LIBCPP_HIDE_FROM_ABI _Alias operator*() const _NOEXCEPT {
       _Alias __val;
-      __builtin_memcpy(&__val, std::__to_address(__base_), sizeof(value_type));
+      __builtin_memcpy(&__val, static_cast<const void*>(std::__to_address(__base_)), sizeof(value_type));
```

Comment:
Just to make it visible again: IMO this is a false-positive, since this is inspecting a non-trivial type, which is perfectly defined behaviour.

Reply:
There's one thing I don't understand: my understanding is that passing a pointer to a non-trivially copyable type to memcpy first or second argument is UB and ` _BaseIter` maybe a non-trivially copyable type, so what happens then?

Reply:
Yeah, this seems like it's plausibly UB. There's an assertion that `__val` is trivial, but I'm not seeing any such assertion for `__base_`; does something else ensure that `_BaseIter` is trivial?

Reply:
@serge-sans-paille Sounds like this requires some extra investigation, which may be holding your patch for longer time than wanted. Would it make more sense to remove all the "silence warning" changes from this patch and focus only on getting the warning in?

Reply:
I'd be happy too, but that would break validation as libcxx has bots running with `-Werror`.

Reply:
Use `#pragma` to disable the warning inline + TODO comment? Not sure if it's allowed by the libcxx guidelines though.

Reply:
Why do you think it is UB? We're only inspecting the memory of the non-trivially copyable type. We're not round-tripping here or anything like that.

Reply:
[cppreference](https://en.cppreference.com/w/cpp/string/byte/memcpy) used to say that it was UB, but now they have changed it to "may be UB":

Honestly, I cannot find any reference in the C++ Standard about this, maybe it's indirectly from [basic.life] of something. The C standard only says:

> The memcpy function copies n characters from the object pointed to by s2 into the object pointed to
by s1. If copying takes place between objects that overlap, the behavior is undefined.

Reply:
Well, cppreference isn't exactly normative. Anyways, where is the actual disconnect? Is it whether you're allowed to
- inspect the bytes of a non-trivially copyable object
- implicitly start the lifetime of a trivial type with bytes from a different object
- do any of the above via `memcpy`
- do the above together in a single `memcpy`
- something else?

Reply:
If we take the perspective of "anything not explicitly defined in the Standard is undefined", then the Standard never specifies the behavior of calling std::memcpy on non-trivially copyable types. It only specifies the behavior for trivially-copyable types in [[basic.types]#2](https://eel.is/c++draft/basic.types.general#2) and [[basic.types]#3](https://eel.is/c++draft/basic.types.general#3).

But if we look at this with a more pragmatic look, I guess the real problem is bypassing the `dst` class constructor, potentially breaking invariants and causing UB due to that. This is perhaps why GCC only warns about the `dst` type being non-trivial. It does **not** warn about the `src` type (which we are discussing here). 

So maybe we could tweak the warning to only warn on the `dst` type instead?

Reply:
Also, [[intro.object]#8.4](https://eel.is/c++draft/intro.object#8.4) says:

> An object of trivially copyable or standard-layout type ([[basic.types.general]](https://eel.is/c++draft/basic.types.general)) shall occupy contiguous bytes of storage[.](https://eel.is/c++draft/intro.object#8.sentence-5)

`std::memcpy` assumes the bytes of the object are laid contiguously in memory, so I guess that's a requirement that also applies to the `src` parameter, not just `dst`.

Not saying we should enforce this in this warning, but this could explain the requirement for TriviallyCopyable.

Reply:
I think the part that gives inspecting an object well-defined behaviour is http://eel.is/c++draft/basic.types#general-4. I'm not 100% sure, though. There might be something more explicit (I was hoping someone here knows). https://eel.is/c++draft/cstring.syn#3 also states that "[The `memcpy` and `memmove`] functions implicitly create objects ([[intro.object]](https://eel.is/c++draft/intro.object)) in the destination region of storage immediately prior to copying the sequence of characters to the destination", which sounds to me a lot like it's allowed to copy the bytes _from_ anywhere.

Reply:
Should we then limit this PR to detecting non-trivially copyable types in the `dst` argument only? Like GCC does.

Reply:
I think so. If this is indeed UB we can still add the warnings later. It's always harder to get users to re-enable a warning than to introduce it in the first place. But let's see what @AaronBallman's thoughts on this are.

Reply:
@AaronBallman IIUC you're saying that it would be UB if some padding bytes ended up in the value representation of a different type, which could indeed arguably be UB. We can't have this scenario here though, since we only instantiate the type with integers as the `_Alias` and make sure that the source objects have a unique object representation via `__is_trivially_equality_comparable`. 

I think it would be interesting to diagnose such cases, but that seems quite complicated, since we'd have to check whether any padding bytes end up in non-padding bytes. (though definitely not impossible)